src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -544,8 +549,14 @@ private ImmutableArray<CompilationEvent> GetPendingEvents_NoLock(ImmutableArray<
         /// <param name="analyzers"></param>
         /// <param name="includeSourceEvents">Indicates if source events (symbol declared, compilation unit completed event) should be included.</param>
         /// <param name="includeNonSourceEvents">Indicates if compilation wide events (compilation started and completed event) should be included.</param>
-        public ImmutableArray<CompilationEvent> GetPendingEvents(ImmutableArray<DiagnosticAnalyzer> analyzers, bool includeSourceEvents, bool includeNonSourceEvents)
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public ImmutableArray<CompilationEvent> GetPendingEvents(
+            ImmutableArray<DiagnosticAnalyzer> analyzers,
+            bool includeSourceEvents,
+            bool includeNonSourceEvents,
+            CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
             lock (_gate)
```

Comment:
I think cancellation check must be inside of the _gate not outside. I think current problem is that, everyone are already at the _gate, waiting to be woke up blocking threads, so, it already passed the cancellation check.

when current one exist the _gate lock after finish running GetPendingEvents_NoLock, next one woke up and again get into GetPendingEvents_NoLock.

I think the check must be right after the lock, not before. it would be even better if check can go inside of the GetPendingEvents_NoLock. but if that method can't be cancelled in the middle due to state management, then check I think at least should be right after _gate not before.

Reply:
Instead of using a normal .net lock, why not use something cancellable, like a SemaphoreSlim?  Woudln't that make things better here?